services/web/server/src/simcore_service_webserver/studies_access.py

```diff
@@ -71,46 +77,40 @@ async def create_temporary_user(request: web.Request):
     email = username + "@guest-at-osparc.io"
     password = get_random_string(min_len=12)
 
-    # creates a user that is marked as ANONYMOUS. see update_user_as_guest
-    user = await db.create_user(
-        {
-            "name": username,
-            "email": email,
-            "password_hash": encrypt_password(password),
-            "status": ACTIVE,
-            "role": ANONYMOUS,
-            "created_ip": get_client_ip(request),
-        }
-    )
-
-    return user
-
-
-async def activate_as_guest_user(user: Dict, app: web.Application):
-    from .login.cfg import get_storage
-    from .login.handlers import GUEST, ANONYMOUS
-    from .resource_manager.websocket_manager import WebsocketRegistry
-
-    # Save time to allow user to be redirected back and not being deleted by GC
-    # SEE https://github.com/ITISFoundation/osparc-simcore/pull/1928#discussion_r517176479
-    EXTRA_TIME_TO_COMPLETE_REDIRECT = 3
-
-    if user.get("role") == ANONYMOUS:
-        db = get_storage(app)
-        username = user["name"]
+    lock_manager: Aioredlock = request.app[APP_CLIENT_REDIS_LOCK_KEY]
 
-        # creates an entry in the socket's registry to avoid garbage collector (GC) deleting it
-        # SEE https://github.com/ITISFoundation/osparc-simcore/issues/1853
+    try:
+        async with await lock_manager.lock(GC_EXECUTION_LOCK, lock_timeout=3):
+            user = await db.create_user(
+                {
+                    "name": username,
+                    "email": email,
+                    "password_hash": encrypt_password(password),
+                    "status": ACTIVE,
+                    "role": GUEST,
+                    "created_ip": get_client_ip(request),
+                }
+            )
+            # time-out lock to prevent GC from deleting this user until it acquires its first resources
+            await lock_manager.lock(
+                GUEST_USER_RC_LOCK_FORMAT.format(user_id=user["id"]),
+                lock_timeout=30,
+            )
+    except LockError as err:
         #
-        registry = WebsocketRegistry(user["id"], f"{username}-guest-session", app)
-        await registry.set_socket_id(
-            f"{username}-guest-socket-id", extra_tll=EXTRA_TIME_TO_COMPLETE_REDIRECT
-        )
+        # Will only attend this request if the GC execution lock can be acquired
+        #
+        # This will prevent disabling GC permanently e.g. due to high frequency guest requests
+        #
+        # At the same time, the lock is timed so in the worst case
+        # this will limit the througput to 1 request every 3 seconds
+        # (provided that nobody has permanently/temporary locked the GC for other reason)
+        #
+        raise web.HTTPTooManyRequests(
+            reason="Number of guests in platform are limited due to available resources"
+        ) from err
```

Comment:
will this be a problem with p2p testing?

Reply:
that's a very good point. Parallel execution of requests probably will not work here. Let me prepare a test that emulate that case and figure out a solution

Reply:
@sanderegg checkout the test I have added in [test_access_to_studies](https://github.com/ITISFoundation/osparc-simcore/pull/1939/files#diff-61854c9acdfaee28412e43c4b3f4f3f5eede80b8487a4e4d3bdc13315f4d486eR393) . There I try to emulate multiple simultaneous requests. The test actually fails with the current implementation. Would you agree that this represents the problem with the p2p testing?

Reply:
yep agree